modules/javafx.base/src/main/java/com/sun/javafx/binding/OldValueCachingListenerList.java

```diff
@@ -0,0 +1,181 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.javafx.binding;
+
+import java.util.Objects;
+
+import javafx.beans.InvalidationListener;
+import javafx.beans.value.ChangeListener;
+import javafx.beans.value.ObservableValue;
+
+/**
+ * Extension of {@link ListenerListBase} which given an {@link ObservableValue}
+ * provides the means to notify all contained listeners with a depth first
+ * approach.<p>
+ *
+ * Note: this listener list stores the latest value for use as the old value when
+ * notifying change listeners. It is recommended to use {@link ListenerList} if
+ * the old value can be provided in some other way as it is more efficient.
+ *
+ * @param <T> the type of the values the observable provides
+ */
+public class OldValueCachingListenerList<T> extends ListenerListBase {
+    private T latestValue;
+
+    /**
+     * This field is only used during notifications, and only relevant
+     * when nested notifications occur. It is used for communicating
+     * information between the different nesting levels. To deeper
+     * nesting levels it contains the number of listeners that have
+     * been notified in higher level loops, while deeper nesting levels
+     * communicate to higher level loops whether a nested notification
+     * actually occurred.<p>
+     *
+     * When its value is zero or positive, it indicates the number of
+     * listeners notified in a higher level loop (minus one), while
+     * when its negative (-1) it indicates to a higher level loop that
+     * a nested notification occurred, requiring, for example, a refresh
+     * of the current value and a new equals check.
+     */
+    private int progress;
+
+    /**
+     * Creates a new instance with two listeners.
+     *
+     * @param listener1 a listener, cannot be {@code null}
+     * @param listener2 a listener, cannot be {@code null}
+     * @throws NullPointerException when any parameter is {@code null}
+     */
+    public OldValueCachingListenerList(Object listener1, Object listener2) {
+        super(listener1, listener2);
+    }
+
+    /**
+     * Returns the latest value stored.
+     *
+     * @return the latest value stored, can be {@code null}
+     */
+    public final T getLatestValue() {
+        return latestValue;
+    }
+
+    /**
+     * Stores the latest value.
+     *
+     * @param value a value to store, can be {@code null}
+     */
+    public final void putLatestValue(T value) {
+        this.latestValue = value;
+    }
+
+    /**
+     * Notifies all listeners using the given observable value as source.
+     *
+     * @param observableValue an {@link ObservableValue}, cannot be {@code null}
+     * @return {@code true} if the listener list is not locked, and it was modified during
+     *     notification otherwise {@code false}
+     */
+    public boolean notifyListeners(ObservableValue<T> observableValue) {
```

Comment:
I agree with you there, and I've been looking what would be a good way to achieve this.  I will take another look soon.  My primary concern is that this is a somewhat critical path, and I would want to ensure that it doesn't cause too much performance regressions (I've already been optimizing all of this code with the help of a JMH test)

Reply:
While looking that code over to see if it could be merged without impacting the general case, I discovered a small bug in the OldValueCaching version.  After I fixed it, the code was even more similar than it was already.  The only different still is the fact that the latest value must be kept track of whenever ObservableValue#getValue is called.

I've now added an extra parameter to the generic version to allow for storing the latest value when it is queried (and not storing it if it's not needed).  This seems to have a minimal performance impact only, so I think the trade off is acceptable.

Reply:
Have you considered adding a method like `void valueUpdated(T value) {}` to `ListenerList`? This will require `ListenerList` to have a type variable `T` (which `OldValueCachingListenerList` adds anyway).

This method could then be called instead of `latestValueTracker.accept(newValue)`, and `OldValueCachingListenerList` can override it and store the value. The advantage of that would be that we don't need the `latestValueTracker` field.

Reply:
Yeah, I'll test that out, it's better not to have that extra field in that class.